### PR TITLE
Server/build jwt auth system basics

### PR DESCRIPTION
**Pull request by travisoverby**

### CODE REVIEW

File: server/config/development.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+db: 
+  host: 'localhost'
+  username: 'postgres'
+  password: 'postgres'
+  synchronize: true
+
+jwt:
+  secret: 'topSecret51'
```
Comment (alizand1992) [inline server/config/development.yml:8]: missing new line

File: server/config/production.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+db:
+  synchronize: false
```
Comment (alizand1992) [inline server/config/production.yml:2]: missing new line

Comment (travisoverby) [reply server/config/production.yml:2]: done

File: server/config/development.yml (hunk 1 above)
Comment (travisoverby) [reply server/config/development.yml:8]: done

File: server/src/app.module.ts [hunk 3]
```diff
@@ -1,10 +1,12 @@
...
+import { AuthModule } from './auth/auth.module';
 
 @Module({
-  imports: [],
-  controllers: [AppController],
-  providers: [AppService],
+  imports: [
+    TypeOrmModule.forRoot(typeOrmConfig),
+    AuthModule,
+  ],
```
Comment (alizand1992) [inline server/src/app.module.ts:10]: Do we need the controllers and all that to be imported?

File: server/src/auth/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            throw new UnauthorizedException('Invalid credentials');
+        }
+
+        const payload: JwtPayload = { username };
+        const accessToken = await this.jwtService.sign(payload);
+        this.logger.debug(`Generated JWT Token with payload ${JSON.stringify(payload)}`);
+
+        return { accessToken };
+    }
+
```
Comment (alizand1992) [inline server/src/auth/auth.service.ts:34]: extra vertical line

File: server/src/auth/dto/auth-credentials.dto.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    @MaxLength(30)
+    @Matches(/^([a-zA-Z0-9_\-\.]+)@([a-zA-Z0-9_\-\.]+)\.([a-zA-Z]{2,5})$/, { message: 'incorrectly formatted email address'} )
+    email: string;
+
+    @IsString()
+    @MinLength(8)
+    @MaxLength(20)
+    @Matches(/((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, { message: 'password is too weak' })
+    password: string;
+
```
Comment (alizand1992) [inline server/src/auth/dto/auth-credentials.dto.ts:15]: extra vertical line

File: server/src/auth/dto/auth-credentials.dto.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+}
```
Comment (alizand1992) [inline server/src/auth/dto/auth-credentials.dto.ts]: missing new line

File: server/src/auth/jwt.strategy.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        const { username } = payload;
+        const user = await this.userRepository.findOne({ username });
+
+        if (!user) {
+            throw new UnauthorizedException();
+        }
+
+        return user;
+    }
+}
```
Comment (alizand1992) [inline server/src/auth/jwt.strategy.ts]: new line

File: server/src/auth/user.entity.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    password: string;
+
+    @Column()
+    salt: string;
+
+    async validatePassword(password: string): Promise<boolean> {
+        const hash = await bcrypt.hash(password, this.salt);
+        return hash === this.password;
+    }
+}
```
Comment (alizand1992) [inline server/src/auth/user.entity.ts]: new line

File: server/src/auth/user.repository.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+            return user.username;
+        }
+
+        return null;
+    }
+
+    private async hashPassword(password: string, salt: string): Promise<string> {
+        return bcrypt.hash(password, salt);
+    }
+}
```
Comment (alizand1992) [inline server/src/auth/user.repository.ts]: new line

File: server/src/app.module.ts (hunk 3 above)
Comment (travisoverby) [reply server/src/app.module.ts]: Maybe later, but it's not necessary at the moment

**Pull request merged**